awx/main/dispatch/worker/callback.py

```diff
@@ -26,89 +30,134 @@ class CallbackBrokerWorker(BaseWorker):
 
     MAX_RETRIES = 2
 
+    def __init__(self):
+        self.buff = {}
+        self.last_write = tz_now()
+
+    def read(self, queue):
+        try:
+            return queue.get(block=True, timeout=.1)
+        except QueueEmpty:
+            return {'event': 'FLUSH'}
+
+    def flush(self, force=False):
+        now = tz_now()
+        if (
+            force or
+            (now - self.last_write).seconds < .1 or
+            any([len(events) == 500 for events in self.buff.values()])
+        ):
+            for cls, events in self.buff.items():
+                logger.debug(f'{cls.__name__}.objects.bulk_create({len(events)})')
+                for e in events:
+                    e.created = e.modified = now
+                try:
+                    cls.objects.bulk_create(events)
+                except IntegrityError:
+                    # this can happen if e.g., the job is deleted *before*
+                    # the event is processed
+                    # if this happens, we have to re-attempt to save the events
+                    # one-by-one, because something in the list is stale
+                    for e in events:
+                        try:
+                            if getattr(e, 'host_id', ''):
+                                # this is one potential IntegrityError we can
+                                # work around - if the host disappears before
+                                # the event can be processed
+                                del e.host_id
+                            cls.objects.bulk_create([e])
+                        except IntegrityError:
+                            logger.exception('Database Error Saving Job Event')
+                for e in events:
+                    emit_event_detail(e)
+            self.buff = {}
+            self.last_write = tz_now()
```

Comment:
Normally, I would encourage calling `now()` early in a function and continue to reference the gotten `now()` value previous. However, this is good that we get the `now()` time again. The `bulk_create` can take a long time. Let's say it takes 1 second because it's doing a very large set of job events. Which means that we are producing a lot of job events. This is all to say, the system is under load. We wouldn't want the `flush()` conditional to trigger more often when under load. We want it to trigger less often. When it triggers less often then we batch more events and we have a better chance of catching up.

By adding this `tz_now()` at the end, we do not penalize (reduce the batch size). The flush logic is purely for better response time and capturing the edge case :+1: